[WPE] Make TouchPoint ID unique

Bastian-Krause opened the pull request:
#### 057b735cbf912f620273106541f2196191c5aa53
<pre>
[WPE] Make TouchPoint ID unique
<a href="https://bugs.webkit.org/show_bug.cgi?id=214865">https://bugs.webkit.org/show_bug.cgi?id=214865</a>

Reviewed by Michael Catanzaro.

The web uses a unique touch point id. A few touch controllers reporting
a unique ID by ABS_MT_TRACKING_ID [1] but libinput and the wayland-protocol
don&apos;t take this information into account. Fix this within WebKit to be
independent of libinput, the wayland-protocol and a wayland-compositor.

This affects TouchEvents in WPE. It will also affect PointerEvents which
will be implemented in a future PR.

Initial patch version by Marco Felsch &lt;m.felsch@pengutronix.de&gt;, tested
and improved by Bastian Krause &lt;bst@pengutronix.de&gt;.

* Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebTouchEvent): Return a WebTouchEvent
with unique WebPlatformTouchPoints by simply incrementing a static id
starting at 2 (1 being WebCore::mousePointerID).

Canonical link: <a href="https://commits.webkit.org/263237@main">https://commits.webkit.org/263237@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/81465c841abddc837aae6e545aa6308bf7b3a431

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3364 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3424 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3544 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4790 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3695 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3490 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3448 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2932 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3405 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3688 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3040 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4608 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1196 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2999 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2925 "4 flakes 139 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2974 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3059 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4352 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3436 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2763 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2999 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3000 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/981 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3004 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3276 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
```diff
@@ -31,6 +31,10 @@
 
+#if ENABLE(TOUCH_EVENTS)
+#include <wtf/HashMap.h>
+#endif
```
Please don't guard includes unless you have to. It just leads to future build failures:

```suggestion
#include <wtf/HashMap.h>
```

Bastian-Krause replied:
Okay.

mcatanzaro commented on Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
```diff
@@ -281,15 +285,24 @@ static WebKit::WebPlatformTouchPoint::TouchPointState stateForTouchPoint(int mai
 WebTouchEvent WebEventFactory::createWebTouchEvent(struct wpe_input_touch_event* event, float deviceScaleFactor)
 {
     auto type = WebEventType::NoType;
+    static HashMap<int32_t, int32_t> activeTrackingTouchPoints;
```
This code is used on Windows so you can't use static objects with destructors because order of destruction is undefined and crashes often ensure.

```suggestion
    static NeverDestroyed<HashMap<int32_t, int32_t>> activeTrackingTouchPoints;
```

You'll have to #include <wtf/NeverDestroyed.h>

Bastian-Krause replied:
Will do, thanks.

mcatanzaro reviewed (changes requested):
I don't actually understand what the touch point ID is, but here's some basic feedback.

mcatanzaro commented on Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
```diff
@@ -282,15 +284,24 @@ static WebKit::WebPlatformTouchPoint::TouchPointState stateForTouchPoint(int mai
+    int32_t pointId;
+
     switch (event->type) {
     case wpe_input_touch_event_type_down:
         type = WebEventType::TouchStart;
+        activeTrackingTouchPoints.get().add(event->id, uniqueTouchPointId);
+        pointId = uniqueTouchPointId;
+        uniqueTouchPointId++;
         break;
     case wpe_input_touch_event_type_motion:
         type = WebEventType::TouchMove;
+        pointId = activeTrackingTouchPoints.get().get(event->id);
         break;
     case wpe_input_touch_event_type_up:
         type = WebEventType::TouchEnd;
+        pointId = activeTrackingTouchPoints.get().take(event->id);
         break;
```
It looks like this only works if the down, motion, and up events all have the exact same `event->id`. So... that is actually true? Seems weird, but OK....

mcatanzaro replied:
I'll land this once you respond here.

Bastian-Krause replied:
> It looks like this only works if the down, motion, and up events all have the exact same `event->id`. So... that is actually true? Seems weird, but OK....

It is true. `wpe_input_touch_event` and its id are based on libinput's seat slot of the touch event ([libinput_event_touch_get_seat_slot()](https://wayland.freedesktop.org/libinput/doc/latest/api/group__event__touch.html)).

The wayland protocol [describes this](https://gitlab.freedesktop.org/wayland/wayland/-/blob/1e259a255a07cd16b31fc99e6bba3e9368fae6d9/protocol/wayland.xml#L2536) as: "This touch point is assigned a unique ID. Future events from this touch point reference this ID. The ID ceases to be valid after a touch up event and may be reused in the future." Compositors implementing the wayland protocol usually just call `libinput_event_touch_get_seat_slot()` themselves.

So the event id is persistent for a touch point (meaning: down, motion [if any], up) for both the DRM platform and wayland.

Bastian-Krause committed:
[WPE] Make TouchPoint ID unique
https://bugs.webkit.org/show_bug.cgi?id=214865

Reviewed by Michael Catanzaro.

The web uses a unique touch point id. A few touch controllers reporting
a unique ID by ABS_MT_TRACKING_ID [1] but libinput and the wayland-protocol
don't take this information into account. Fix this within WebKit to be
independent of libinput, the wayland-protocol and a wayland-compositor.

This affects TouchEvents in WPE. It will also affect PointerEvents which
will be implemented in a future PR.

Initial patch version by Marco Felsch <m.felsch@pengutronix.de>, tested
and improved by Bastian Krause <bst@pengutronix.de>.

* Source/WebKit/Shared/libwpe/WebEventFactory.cpp:
(WebKit::WebEventFactory::createWebTouchEvent): Return a WebTouchEvent
with unique WebPlatformTouchPoints by simply incrementing a static id
starting at 2 (1 being WebCore::mousePointerID).

Canonical link: https://commits.webkit.org/263237@main
